eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -503,6 +505,43 @@ class
 
   expectTxSuccess :: HasCallStack => Tx era -> ImpTestM era ()
 
+impSatisfySignature ::
+  KeyHash 'Witness ->
+  Set.Set (KeyHash 'Witness) ->
+  ImpTestM era (Maybe (Map (KeyHash 'Witness) (KeyPair 'Witness)))
+impSatisfySignature keyHash providedVKeyHashes = do
+  keyPairs <- gets impKeyPairs
+  if keyHash `Set.member` providedVKeyHashes
+    then
+      pure $ Just mempty
+    else
+      pure $ Map.singleton keyHash <$> Map.lookup keyHash keyPairs
+
+impSatisfyMNativeScripts ::
+  ShelleyEraImp era =>
+  Int ->
+  -- | Nr of scripts to satisfy
+  StrictSeq (NativeScript era) ->
+  -- | List of scripts to satisfy a subset of
+  Set.Set (KeyHash 'Witness) ->
+  -- | Set of Witnesses that have already been satisfied
+  TxBody era ->
+  -- | The transaction body that the scripts wil lbe applied to
+  ImpTestM era (Maybe (Map (KeyHash 'Witness) (KeyPair 'Witness)))
+impSatisfyMNativeScripts nrOfScriptsToSatisfy scripts providedVKeyHashes txBody =
+  go nrOfScriptsToSatisfy scripts
+  where
+    go m Empty
+      | m <= 0 = pure $ Just mempty
+      | otherwise = pure Nothing
+    go m (x :<| xs) = do
+      satisifed <- impSatisfyNativeScript providedVKeyHashes txBody x
+      case satisifed of
+        Nothing -> go m xs
+        Just kps -> do
+          kps' <- go (m - 1) xs
+          pure $ (kps <>) <$> kps'
```

Comment:
You can leave it as it is too. I don't know if it makes a difference.
```suggestion
          fmap (kps <>) <$> go (m - 1) xs
```

Reply:
That's great, thanks!